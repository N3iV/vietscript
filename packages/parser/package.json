{
  "name": "@vietscript/parser",
  "version": "0.0.1",
  "description": "Vietnamese programming language - Ngôn ngữ lập trình tiếng Việt.",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "types": "./dist/types/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "typecheck": "tsc --noEmit",
    "build": "rimraf ./dist && vite build && cp ./dist/index.umd.cjs ./public/index.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/imrim12/vietscript.git",
    "directory": "packages/parser"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "keywords": [
    "language",
    "vietscript",
    "vietnamese",
    "programming language"
  ],
  "author": "imrim12",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/imrim12/vietscript/issues"
  },
  "homepage": "https://github.com/imrim12/vietscript#readme",
  "devDependencies": {
    "@vietscript/shared": "workspace:*",
    "@types/node": "^18.14.1",
    "@vitest/ui": "^0.24.5",
    "lodash": "^4.17.21",
    "rimraf": "^3.0.2",
    "typescript": "^4.9.4",
    "vite": "^4.0.4",
    "vite-plugin-dts": "^1.6.6",
    "vitest": "^0.28.3"
  }
}