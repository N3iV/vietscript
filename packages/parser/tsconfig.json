{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "outDir": "./dist",
    "types": [
      "@types/node",
      "vite/client",
      "vitest/globals"
    ],
    "paths": {
      "@parser": [
        "./src/index.ts"
      ],
      "@parser/*": [
        "./src/*"
      ]
    }
  },
  "include": [
    "./**/*.ts"
  ],
  "exclude": [
    "./node_modules",
    "./dist"
  ]
}