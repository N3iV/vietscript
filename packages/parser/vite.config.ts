import path from "node:path";

import dts from "vite-plugin-dts";
import { defineConfig } from "vite";

export default defineConfig({
  resolve: {
    alias: {
      "@parser": path.resolve("./src"),
    },
  },
  build: {
    emptyOutDir: false,
    lib: {
      name: "@vietscript/parser",
      entry: path.resolve("./src/index.ts"),
      formats: ["es", "cjs", "umd"],
      fileName: "index",
    },
    rollupOptions: {
      external: [],

      output: {
        exports: "named",
        globals: {},
      },
    },
  },
  plugins: [
    dts({
      root: ".",
      entryRoot: "./src",
      outputDir: "./dist/types",
    }),
  ],
});
