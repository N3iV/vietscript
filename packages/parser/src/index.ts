import { Parser } from "./parser";

const parser = new Parser();

export default parser;

export { Tokenizer } from "./tokenizer";
export { Parser } from "./parser";

if (typeof window !== "undefined") {
  customElements.define(
    "vi-script",
    class extends HTMLElement {
      constructor() {
        super();
      }

      connectedCallback() {
        setTimeout(() => {
          const script = this.childNodes.item(0).textContent?.trim();

          this.innerHTML = "";

          if (script) {
            // executor.execute(script);
          }
        });
      }
    },
  );

  // @ts-ignore
  window.VietScript = { parser };
}
