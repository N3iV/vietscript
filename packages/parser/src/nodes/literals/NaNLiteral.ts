import { Node } from "@vietscript/shared";
import { Parser } from "@parser/parser";

export class NaNLiteral implements Node {
  type = "NaNLiteral";

  value: number;

  raw: string;

  start: number;

  end: number;

  constructor(parser: Parser) {
    const token = parser.eat("NaN");

    this.start = token.start;
    this.end = token.end;

    this.value = Number.NaN;
    this.raw = "NaN";
  }
}
