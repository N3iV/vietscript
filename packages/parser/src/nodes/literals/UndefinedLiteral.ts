import { Node } from "@vietscript/shared";
import { Parser } from "@parser/parser";

export class UndefinedLiteral implements Node {
  type = "UndefinedLiteral";

  value: undefined;

  raw: string;

  start: number;

  end: number;

  constructor(parser: Parser) {
    const token = parser.eat("Undefined");

    this.start = token.start;
    this.end = token.end;

    // TODO: toPlainObject removes the undefined value, lead to fail test
    this.value = undefined;
    this.raw = "undefined";
  }
}
