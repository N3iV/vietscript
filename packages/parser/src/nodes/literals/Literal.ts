import { Parser } from "@lang/parser";

import { BooleanLiteral } from "./BooleanLiteral";
import { NaNLiteral } from "./NaNLiteral";
import { NullLiteral } from "./NullLiteral";
import { NumericLiteral } from "./NumericLiteral";
import { StringLiteral } from "./StringLiteral";
import { UndefinedLiteral } from "./UndefinedLiteral";

export class Literal {
  type = "Literal";

  value: number | boolean | string | null | undefined;

  raw: string;

  constructor(parser: Parser) {
    let literal: any;

    switch (parser.lookahead?.type) {
      case "Number": {
        literal = new NumericLiteral(parser);
        break;
      }
      case "String": {
        literal = new StringLiteral(parser);
        break;
      }
      case "Boolean": {
        literal = new BooleanLiteral(parser);
        break;
      }
      case "NaN": {
        literal = new NaNLiteral(parser);
        break;
      }
      case "Null": {
        literal = new NullLiteral(parser);
        break;
      }
      case "Undefined": {
        literal = new UndefinedLiteral(parser);
        break;
      }
    }

    this.value = literal.value;
    this.raw = literal.raw;
  }
}
