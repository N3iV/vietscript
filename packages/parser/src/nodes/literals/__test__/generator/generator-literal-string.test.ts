import { Literal } from "@parser/nodes/literals/Literal";
import generate from "@babel/generator";

import parser from "../../../../setup-test";

describe("generator-literal-string.test", () => {
  it("should generate the javascript from the syntax normally", () => {
    const code = `"Chào thế giới!"`;
    const ast = parser.parse(code, Literal);

    const result = generate(ast);

    expect(result.code).toMatchSnapshot();
  });
});
