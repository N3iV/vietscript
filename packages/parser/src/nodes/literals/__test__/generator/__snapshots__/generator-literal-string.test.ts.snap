// Vitest Snapshot v1

exports[`generator-literal-string.test > should generate the javascript from the syntax normally 1`] = `"\\"Chào thế giới!\\""`;
